### ISSUE TITLE
Feature: Define a value that will trigger module defaults

### ISSUE THREAD

**Issue by flmmartins** [hashicorp/terraform#29517]
### Current Terraform Version

```
1.0.2
```

### Use-cases

Let say we have a loop in a module with count,

```terraform
module "acr" {
  source = "../../myacrmodule"
  count   = length(var.acr)
  name                                                 = var.acr["name"]
  sku_tier                                             = try(var.acr["sku_tier"], "Basic")
  admin_enabled                                 = try(var.acr["admin_enabled"], false)
  network_rule_default_action          = try(var.acr["network_rule_default_action"], "Deny")
  whitelisted_cidrs                              = try(var.acr["whitelisted_cidrs"], [])
  whitelisted_subnets                         = try(var.acr["whitelisted_subnets"], [])
  retention_policy_enabled                = try(var.acr["retention_policy_enabled"], false)
  retention_policy_number_of_days = try(var.acr["retention_policy_number_of_days"], 7)
  georeplications                                 = try(var.acr["georeplications"], [])
  resource_group = {
    name     = azurerm_resource_group.main[var.regions[0]].name
    location = azurerm_resource_group.main[var.regions[0]].location
  }
  tags = merge(local.tags, {})
}
```

The ACR module in this case have optional keys so I try to fetch the key and if it's not there I use a default but the default is already inside the module itself. So the problem here is why I need to define this 2 times? In the module AND in the module call?

This situation happen when we use the same code with different variables where they might have or not have a number of resources. For the example above prd.tfvars could have an ACR and dev.tfvars could not....

Another use case:

```terraform

locals {
  groups = defaults(var.groups, {
    group_description             = ""
    subgroup_creation_level  = "owner"
    project_creation_level      = "noone"
    group_visibility                  = "private"
    request_access_enabled = true
    group_variables                = {}
    repositories                       = {}
  })
  
module "groups-l1" {
  source = "../../mygit"

  for_each = toset(local.groups_l1)

  group_path                        = each.key
  group_name                     = local.groups[each.key].display_name
  parent_id                          = local.top_group_id
  group_description            = local.groups[each.key].group_description
  group_visibility                 = local.groups[each.key].group_visibility
  subgroup_creation_level = local.groups[each.key].subgroup_creation_level
  project_creation_level      = local.groups[each.key].project_creation_level
  request_access_enabled = local.groups[each.key].request_access_enabled
  group_variables                = local.groups[each.key].group_variables
  group_members               = local.groups[each.key].group_members
  repositories                       = local.groups[each.key].repositories
  gitlab_token                      = var.git_token
}
```

In this example we have an Optional object.. but hey we need to define defaults again although they are already in the module.

In this example I could also have a tfvars and that defines a group that has these attributes or not.... 

### Attempted Solutions

The solution for this is always define a default 2 times: one when you call the module and then inside module itself which causes duplicated code. Basically all possible solutions for this will include you defining the defaults 2x

### Proposal

Today we use defaults for modules when we omit the whole parameter (key and value). It would be nice if I can say: if this value is NULL then automatically use the module defaults.

So in the first USE case we could eliminate the try and if the map has the value, use it. If not that is NULL then the module defaults would be used.

Same thing with Optional Objects. They already have default as null so  in the second example I would just call the object without having to define local (and defaults) and then this would activate module defaults.

### References

- https://github.com/hashicorp/terraform/issues/29488
- https://github.com/hashicorp/terraform/issues/24142

**Comment by EronWright** [comment]
A possible solution has been contributed to TF 1.1:
https://github.com/hashicorp/terraform/pull/29832

**Comment by korinne** [comment]
Hi all, we recently released the [Terraform v1.3 alpha](https://releases.hashicorp.com/terraform/1.3.0-alpha20220608), which includes the ability to mark object type attributes as optional and set default values.

I'm following up with issues that have provided feedback on the previous experiment, and wanted to invite you all to try the alpha and provide any feedback on this updated design. You can learn more/add comments[ here](https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550). Thank you so much in advance, and hope you like the feature!

**Comment by apparentlymart** [comment]
Hi again!

As @EronWright noted above, we implemented a solution to this in #29832, which was release a while ago but we didn't notice at the time that this issue was affected by it.

To preserve backward compatibility with existing modules, the new behavior requires explicitly setting `nullable = false` in the variable declaration. If you do so then Terraform will guarantee that the input variable can never be null, which it will achieve by using your specified default value instead if the caller sets the value to null. We will consider making this behavior the default in a future edition of the Terraform language, but we don't have any immediate plans to do that.

Thanks!
